<?xml version="1.0"?>
<robot name="gz_arm">

  <material name="red">
    <color rgba="1 0 0 0.7"/>
  </material>
  <material name="green">
    <color rgba="0 1 0 0.7"/>
  </material>
  <material name="blue">
    <color rgba="0 0 1 0.7"/>
  </material>
  <material name="magenta">
    <color rgba="0.7 0 0.7 0.7"/>
  </material>
  <material name="grey">
    <color rgba="0.7 0.7 0.7 0.7"/>
  </material>
  <material name="yellow">
    <color rgba="0.7 0.7 0 0.7"/>
  </material>

  
  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.2 0.2 0.1"/>
      </geometry>
      <origin xyz="0 0 0.05" rpy="0 0 0"/>
      <material name="red"/>
    </visual>
  </link>

  <link name="arm1_link">
    <visual name='link'>
      <geometry>
        <box size="0.3 0.04 0.03"/>
      </geometry>
      <origin xyz="0.15 0 0" rpy="0 0 0"/>
      <material name="green"/>
    </visual>
    <visual name='joint1'>
      <geometry>
        <cylinder radius="0.025" length="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="green"/>
    </visual>
    <visual name='joint2'>
      <geometry>
        <cylinder radius="0.025" length="0.05"/>
      </geometry>
      <origin xyz="0.3 0 0" rpy="0 0 0"/>
      <material name="green"/>
    </visual>
  </link>

  <link name="arm2_link">
    <visual name='link'>
      <geometry>
        <box size="0.2 0.04 0.03"/>
      </geometry>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <material name="blue"/>
    </visual>
    <visual name='joint1'>
      <geometry>
        <cylinder radius="0.025" length="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="blue"/>
    </visual>
    <visual name='joint2'>
      <geometry>
        <cylinder radius="0.025" length="0.05"/>
      </geometry>
      <origin xyz="0.2 0 0" rpy="0 0 0"/>
      <material name="blue"/>
    </visual>
  </link>

  <link name="arm3_link">
    <visual name='link'>
      <geometry>
        <box size="0.15 0.04 0.03"/>
      </geometry>
      <origin xyz="0.075 0 0" rpy="0 0 0"/>
      <material name="magenta"/>
    </visual>
    <visual name='joint1'>
      <geometry>
        <cylinder radius="0.025" length="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="magenta"/>
    </visual>
    <visual name='joint2'>
      <geometry>
        <cylinder radius="0.025" length="0.05"/>
      </geometry>
      <origin xyz="0.15 0 0" rpy="0 0 0"/>
      <material name="magenta"/>
    </visual>
  </link>

  <link name="camera_link">
    <visual name='link'>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 1.5707963267948966 0"/>
      <material name="yellow"/>
    </visual>
  </link>
  
  <joint name="base_to_arm1_joint" type="revolute">
    <parent link="base_link"/>
    <child link="arm1_link"/>
    <limit effort="100.0" lower="-3.0" upper="3.0" velocity="0.5"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0 0 0.125" rpy="0 0 0"/>
  </joint>

  <joint name="arm1_to_arm2_joint" type="revolute">
    <parent link="arm1_link"/>
    <child link="arm2_link"/>
    <limit effort="100.0" lower="-3.0" upper="3.0" velocity="0.5"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0.3 0 0.05" rpy="0 0 0"/>
  </joint>

  <joint name="arm2_to_arm3_joint" type="revolute">
    <parent link="arm2_link"/>
    <child link="arm3_link"/>
    <limit effort="100.0" lower="-3.0" upper="3.0" velocity="0.5"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0.2 0 0.05" rpy="0 0 0"/>
  </joint>

  <joint name="arm3_to_camera_joint" type="fixed">
    <parent link="arm3_link"/>
    <child link="camera_link"/>
    <origin xyz="0.15 0 0.05" rpy="-1.5707963267948966 0 -1.5707963267948966"/>
  </joint>
  
  <!-- velocity -->
  <ros2_control name="GazeboSimSystem" type="system">
    <hardware>
      <plugin>gz_ros2_control/GazeboSimSystem</plugin>
    </hardware>
    <joint name="base_to_arm1_joint">
      <command_interface name="position">
        <param name="min">-3</param>
        <param name="max">3</param>
      </command_interface>
      <command_interface name="velocity">
        <param name="min">-1</param>
        <param name="max">1</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">1.0</param>
      </state_interface>
      <state_interface name="velocity"/>
    </joint>
    <joint name="arm1_to_arm2_joint">
      <command_interface name="position">
        <param name="min">-3</param>
        <param name="max">3</param>
      </command_interface>
      <command_interface name="velocity">
        <param name="min">-1</param>
        <param name="max">1</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">-2.0</param>
      </state_interface>
      <state_interface name="velocity"/>
    </joint>
    <joint name="arm2_to_arm3_joint">
      <command_interface name="position">
        <param name="min">-3</param>
        <param name="max">3</param>
      </command_interface>
      <command_interface name="velocity">
        <param name="min">-1</param>
        <param name="max">1</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">1.0</param>
      </state_interface>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>

</robot>
